### ISSUE TITLE
Remaining tasks for CHARMM conversion

### ISSUE THREAD

**Issue by jchodera** [pandegroup/openmm#1366]
We have two major remaining tasks for the CHARMM36 conversion:

1.**Identify a way to let parameters in one `ffxml` file override those in another file.** We asked the CHARMM developers, but even they don't know which one is right, so it would be difficult (if not impossible) for us to order them in terms of priority.  The CHARMM developers will revisit the conflicts, and may eventually produce a consistent set of files, but I think we need to implement some mode for OpenMM to replicate the last-one-wins behavior. Otherwise, the CHARMM files may never make it in. I think placing the burden of selecting the right file order on the user (with appropriate warnings!) is a reasonable alternative to dropping support for having access to CHARMM parameters via `ForceField`. (cc: #1344 #1281)
2. **Identify a strategy for handling CHARMM patches.** The simple solution to this one is less clear, and may merit a quick video chat to discuss possible strategies. (cc: #1335)

**Comment by swails** [comment]
What about releasing a monolithic force field XML that makes the vast majority of the FF people would want available and punt on the weird corner cases until CHARMM comes up with a real solution?

**Comment by swails** [comment]
I mean if we compare to what we do with Amber, OpenMM makes only a tiny fraction of the Amber FF available to users (no gaff or antechamber, no GLYCAM, no Lipid FF, ... etc.).  And the quality and ease-of-use of GAFF/antechamber (over most of the alternatives that I know of) helped win Amber a lot of users.  The end-goal here is to make all of it available (and your group is putting a lot of effort into that), but it doesn't seem like you guys had the mind-set of "we must support all aspects of the Amber force field the first time we release any of it." (Amber FFs were converted/released well before I got involved.)

Let's focus on releasing the core components that many people care about (proteins, nucleic acids, maybe lipids and carbohydrates), and be patient with the weird quirky parts that not even CHARMM knows what to do with until they get ironed out.  At least that's my take on things.

**Comment by jchodera** [comment]
> What about releasing a monolithic force field XML that makes the vast majority of the FF people would want available and punt on the weird corner cases until CHARMM comes up with a real solution?

1. There may never be a real solution.
2. Without a way to override things, we are essentially locking people into a world they may not want to live in.  We might be serving an audience of measure zero.
3. *We* don't even know what files people might reasonably want to use together, and figuring out the largest non-conflicting set is a nontrivial graph theoretic problem.

TL;DR: It doesn't make anything easier for us or other people, so let's not do that.

**Comment by jchodera** [comment]
> Let's focus on releasing the core components that many people care about (proteins, nucleic acids, maybe lipids and carbohydrates), and be patient with the weird quirky parts that not even CHARMM knows what to do with until they get ironed out. At least that's my take on things.

I disagree.  If we convert *everything* and have a simple "last one wins" mechanism (either by default or that people can enable) along with a single file that includes the basics (proteins and nucleic acids), we still let people do more complicated things by just knowing enough about CHARMM to know what they want to list in the arguments to `ForceField.xml`, and many people can get by with that.  It's the least amount of work to enable the most amount of science.

The selfish reason here is that "just proteins and nucleic acids" and even some lipids and carbohydrates wouldn't even work for what we want to use CHARMM for.  There is really zero motivation here to spend a lot of time on a solution we can't even use ourselves.

**Comment by peastman** [comment]
Can this be closed?  I think we've made the decisions on both of these and already implemented them.

**Comment by swails** [comment]
Still waiting on #1453 to be merged, yes?  Also, #1399 is part of this, too.

**Comment by jchodera** [comment]
CHARMM conversion is complete!
